File: typescript/tests/integration/shutdown-unreachable-endpoint.test.ts
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
// @vitest-environment node

/**
 * Regression: shutdownTelemetry must not hang when the OTLP logs endpoint
 * is unreachable.
 *
 * Mirrors the Python regression at
 * tests/regression/test_shutdown_unreachable_endpoint.py. The Python side
 * fixed the same class of bug (BatchLogRecordProcessor.force_flush blocking
 * indefinitely on an unreachable collector). TypeScript's shutdownTelemetry
 * previously awaited `Promise.allSettled(providers.map(p => p.forceFlush()))`
 * with no per-promise timeout, so a hanging forceFlush would hang the whole
 * shutdownTelemetry call. The fix wraps each forceFlush+shutdown pair with
 * a deadline derived from `PROVIDE_EXPORTER_LOGS_SHUTDOWN_TIMEOUT_SECONDS`.
 */

import { createServer, type Server } from 'node:net';
import { afterEach, beforeEach, describe, expect, it } from 'vitest';

import {
  getConfig,
  getLogger,
  registerOtelProviders,
  resetTelemetryState,
  setupTelemetry,
  shutdownTelemetry,
} from '../../src/index.js';

async function reserveClosedPort(): Promise<number> {
  // Bind a TCP port, capture its number, then close so connects fail fast
  // with ECONNREFUSED (instead of timing out at the OS level, which would
  // mask whether our deadline actually fired).
  return new Promise<number>((resolve, reject) => {
    const server: Server = createServer();
    server.unref();
    server.listen(0, '127.0.0.1', () => {
      const addr = server.address();
      if (addr === null || typeof addr === 'string') {
        server.close();
        reject(new Error('failed to allocate a localhost port'));
        return;
      }
      const port = addr.port;
      server.close(() => resolve(port));
    });
  });
}

describe('shutdownTelemetry — unreachable OTLP logs endpoint', () => {
  beforeEach(() => {
    resetTelemetryState();
  });
  afterEach(async () => {
    await shutdownTelemetry();
    resetTelemetryState();
    delete process.env['OTEL_EXPORTER_OTLP_ENDPOINT'];
    delete process.env['OTEL_EXPORTER_OTLP_LOGS_ENDPOINT'];
    delete process.env['PROVIDE_EXPORTER_LOGS_SHUTDOWN_TIMEOUT_SECONDS'];
    delete process.env['PROVIDE_LOG_OTLP_ENABLED'];
  });

  it('returns within the deadline when the logs endpoint refuses connections', async () => {
    const port = await reserveClosedPort();
    const endpoint = `http://127.0.0.1:${port}`;

    setupTelemetry({
      serviceName: 'shutdown-unreachable-regression',
      otelEnabled: true,
      tracingEnabled: false,
      metricsEnabled: false,
      otlpEndpoint: endpoint,
      // Aggressive deadline so a regression in the bounded shutdown is
      // immediately visible. Pre-fix this would hang indefinitely; post-fix
      // the deadline forces shutdownTelemetry to return.
      exporterLogsShutdownTimeoutMs: 250,
      consoleOutput: false,
      captureToWindow: false,
    });
    await registerOtelProviders(getConfig());

    // Emit a log so the BatchLogRecordProcessor has something to flush;
    // an empty queue would let force_flush resolve instantly even without
    // the fix.
    getLogger('shutdown-unreachable-regression').info({
      event: 'shutdown_unreachable_probe',
    });

    const started = performance.now();
    await shutdownTelemetry();
    const elapsed = performance.now() - started;

    // 250ms deadline + node + GC noise. Pre-fix this would never return
    // within any bound because Promise.allSettled awaits the OTel exporter's
    // own retry chain.
    expect(elapsed).toBeLessThan(2000);
  });

  it('PROVIDE_LOG_OTLP_ENABLED=false avoids the OTLP log path entirely', async () => {
    const port = await reserveClosedPort();
    const endpoint = `http://127.0.0.1:${port}`;

    setupTelemetry({
      serviceName: 'shutdown-unreachable-disabled-regression',
      otelEnabled: true,
      tracingEnabled: false,
      metricsEnabled: false,
      otlpEndpoint: endpoint,
      // Disable the OTLP log provider — when otlpLogsEnabled=false the
      // SDK is never installed, so there's nothing to flush.
      otlpLogsEnabled: false,
      // Use the default deadline; the test asserts the path is fast on its
      // own, not because of the deadline.
      consoleOutput: false,
      captureToWindow: false,
    });
    await registerOtelProviders(getConfig());

    getLogger('shutdown-unreachable-disabled-regression').info({
      event: 'shutdown_unreachable_disabled_probe',
    });

    const started = performance.now();
    await shutdownTelemetry();
    const elapsed = performance.now() - started;

    // No OTLP log provider was registered, so flush is trivially fast.
    expect(elapsed).toBeLessThan(1000);
  });
});

File: rust/tests/shutdown_unreachable_endpoint_test.rs
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0
// SPDX-Comment: Part of provide-telemetry.
//
//! Regression: `shutdown_telemetry()` must return promptly when the OTLP
//! log endpoint is unreachable.
//!
//! Mirrors the Python/TS/Go regressions. Without the bounded-shutdown
//! deadline (PROVIDE_EXPORTER_LOGS_SHUTDOWN_TIMEOUT_SECONDS) the OTel SDK's
//! `LoggerProvider::force_flush`/`shutdown` can sit in its internal retry
//! loop indefinitely against a closed-port collector.

#![cfg(feature = "otel")]

use std::net::TcpListener;
use std::time::{Duration, Instant};

use provide_telemetry::testing::acquire_test_state_lock;
use provide_telemetry::{setup_telemetry, shutdown_telemetry};

fn reserve_closed_port() -> u16 {
    let listener = TcpListener::bind("127.0.0.1:0").expect("listen on 127.0.0.1:0");
    let port = listener
        .local_addr()
        .expect("local_addr should succeed")
        .port();
    drop(listener); // close, so connects refuse instantly
    port
}

fn restore_var(key: &str, previous: Option<String>) {
    match previous {
        Some(value) => std::env::set_var(key, value),
        None => std::env::remove_var(key),
    }
}

#[test]
fn shutdown_telemetry_returns_within_deadline_with_unreachable_endpoint() {
    let _guard = acquire_test_state_lock();
    let _ = shutdown_telemetry();
    provide_telemetry::otel::_reset_otel_for_tests();

    let endpoint = format!("http://127.0.0.1:{}/", reserve_closed_port());
    let endpoint_key = "OTEL_EXPORTER_OTLP_ENDPOINT";
    let shutdown_key = "PROVIDE_EXPORTER_LOGS_SHUTDOWN_TIMEOUT_SECONDS";
    let trace_key = "PROVIDE_TRACE_ENABLED";
    let metrics_key = "PROVIDE_METRICS_ENABLED";

    let prev_endpoint = std::env::var(endpoint_key).ok();
    let prev_shutdown = std::env::var(shutdown_key).ok();
    let prev_trace = std::env::var(trace_key).ok();
    let prev_metrics = std::env::var(metrics_key).ok();

    std::env::set_var(endpoint_key, &endpoint);
    // 250ms deadline: pre-fix the OTel SDK's batch processor flush could
    // sit on the unreachable endpoint until its per-export timeout fired
    // (10s default). 250ms gives the bounded path room to fire and the
    // overall test wall time stays well under 2s even with noise.
    std::env::set_var(shutdown_key, "0.25");
    // Disable trace+metrics so only the logs OTLP path is exercised;
    // mirrors the cross-language pattern that proves disabling those is
    // insufficient on its own.
    std::env::set_var(trace_key, "false");
    std::env::set_var(metrics_key, "false");

    setup_telemetry().expect("setup should succeed even with unreachable endpoint");

    let started = Instant::now();
    shutdown_telemetry().expect("shutdown should return cleanly");
    let elapsed = started.elapsed();

    provide_telemetry::otel::_reset_otel_for_tests();
    restore_var(endpoint_key, prev_endpoint);
    restore_var(shutdown_key, prev_shutdown);
    restore_var(trace_key, prev_trace);
    restore_var(metrics_key, prev_metrics);

    // 250ms deadline + scheduling noise. A regression in the bounded
    // helper would push this well past the threshold.
    assert!(
        elapsed < Duration::from_secs(2),
        "shutdown_telemetry took {elapsed:?} with unreachable endpoint, expected <2s",
    );
}

#[test]
fn disable_log_otlp_avoids_provider_install_and_keeps_shutdown_fast() {
    let _guard = acquire_test_state_lock();
    let _ = shutdown_telemetry();
    provide_telemetry::otel::_reset_otel_for_tests();

    let endpoint = format!("http://127.0.0.1:{}/", reserve_closed_port());
    let endpoint_key = "OTEL_EXPORTER_OTLP_ENDPOINT";
    let enabled_key = "PROVIDE_LOG_OTLP_ENABLED";
    let trace_key = "PROVIDE_TRACE_ENABLED";
    let metrics_key = "PROVIDE_METRICS_ENABLED";

    let prev_endpoint = std::env::var(endpoint_key).ok();
    let prev_enabled = std::env::var(enabled_key).ok();
    let prev_trace = std::env::var(trace_key).ok();
    let prev_metrics = std::env::var(metrics_key).ok();

    std::env::set_var(endpoint_key, &endpoint);
    std::env::set_var(enabled_key, "false");
    std::env::set_var(trace_key, "false");
    std::env::set_var(metrics_key, "false");

    setup_telemetry().expect("setup should succeed with otlp_enabled=false");

    let started = Instant::now();
    shutdown_telemetry().expect("shutdown should return cleanly");
    let elapsed = started.elapsed();

    provide_telemetry::otel::_reset_otel_for_tests();
    restore_var(endpoint_key, prev_endpoint);
    restore_var(enabled_key, prev_enabled);
    restore_var(trace_key, prev_trace);
    restore_var(metrics_key, prev_metrics);

    assert!(
        elapsed < Duration::from_secs(1),
        "shutdown_telemetry took {elapsed:?} with OTLP logs disabled, expected <1s",
    );
}
